pkg/check/check.go

```diff
@@ -0,0 +1,128 @@
+package check
+
+import (
+	"context"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+)
+
+// Checker is interface that defines checker to check configurations of system.
+// It is mainly used for configuration checking of data synchronization between database systems.
+type Checker interface {
+	Name() string
+	Check(ctx context.Context) *Result
+}
+
+// State is state of check
+type State string
+
+const (
+	// StateSuccess indicates that the check was successful
+	StateSuccess State = "success"
+	// StateFailure indicates that the check was faiful
+	StateFailure State = "fail"
+	// StateWarning indicates that the check had warnings
+	StateWarning State = "warn"
+)
+
+// Result is result of check
+type Result struct {
+	ID          uint64 `json:"id"`
+	Name        string `json:"name"`
+	Desc        string `json:"desc"`
+	State       State  `json:"state"`
+	ErrorMsg    string `json:"errorMsg"`
+	Instruction string `json:"instruction"`
+	Extra       string `json:"extra"`
+}
+
+// ResultSummary is summary of all check results
+type ResultSummary struct {
+	Passed     bool  `json:"passed"`
+	Total      int64 `json:"total"`
+	Successful int64 `json:"successful"`
+	Failed     int64 `json:"failed"`
+	Warning    int64 `json:"warning"`
+}
+
+// Results contains all check results and summary
+type Results struct {
+	Results []*Result      `json:"results"`
+	Summary *ResultSummary `json:"summary"`
+}
+
+// Do executes several checkers.
+func Do(ctx context.Context, checkers []Checker) (*Results, error) {
+	results := &Results{
+		Results: make([]*Result, 0, len(checkers)),
+	}
+	if len(checkers) == 0 {
+		results.Summary = &ResultSummary{Passed: true}
+		return results, nil
+	}
+
+	var (
+		wg         sync.WaitGroup
+		finished   bool
+		total      int64
+		successful int64
+		failed     int64
+		warning    int64
+		err        error
+	)
+	total = int64(len(checkers))
+
+	resultCh := make(chan *Result)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case result := <-resultCh:
```

Comment:
how about `for result := range resultCh` ?

Reply:
both are ok

Reply:
I know, maybe `for result := range resultCh` seems simpler, maybe

Reply:
I mean If something can be changed or not, leave it to your favor :smile: